pkg/storage/concurrency/concurrency_manager.go

```diff
@@ -0,0 +1,363 @@
+		// lock table to allow contending transactions to proceed.
+		// for _, arg := range req.Requests {
+		// 	switch t := arg.GetInner().(type) {
+		// 	case *roachpb.ResolveIntentRequest:
+		// 		_ = t
+		// 	case *roachpb.ResolveIntentRangeRequest:
+		// 		_ = t
+		// 	}
+		// }
+	}
+	return nil, nil
+}
+
+// shouldAcquireLatches determines whether the request should acquire latches
+// before proceeding to evaluate. Latches are used to synchronize with other
+// conflicting requests, based on the Spans collected for the request. Most
+// request types will want to acquire latches.
+func shouldAcquireLatches(req Request) bool {
+	switch {
+	case req.ReadConsistency != roachpb.CONSISTENT:
+		// Only acquire latches for consistent operations.
+		return false
+	case req.isSingle(roachpb.RequestLease):
+		// Don't acquire latches for lease requests. These are run on replicas
+		// that do not hold the lease, so acquiring latches wouldn't help
+		// synchronize with other requests.
+		return false
+	}
+	return true
+}
+
+// shouldWaitOnConflicts determines whether the request should wait on locks and
+// wait-queues owned by other transactions before proceeding to evaluate. Most
+// requests will want to wait on conflicting transactions to ensure that they
+// are sufficiently isolated during their evaluation, but some "isolation aware"
+// requests want to proceed to evaluation even in the presence of conflicts
+// because they know how to handle them.
+func shouldWaitOnConflicts(req Request) bool {
+	// TODO what's the best way to define this. There are request types that do
+	// want to wait of locks like (PutRequest and ScanRequest) and then there
+	// are those that don't want to wait on locks like (QueryIntentRequest,
+	// RefreshRequest, and ResolveIntentRequest). Should we define a flag for
+	// this?
+	should := false
+	for _, ru := range req.Requests {
+		arg := ru.GetInner()
+		if roachpb.IsTransactional(arg) {
+			switch arg.Method() {
+			case roachpb.HeartbeatTxn:
+			case roachpb.Refresh:
+			case roachpb.RefreshRange:
+			default:
+				should = true
+			}
+		}
+	}
+	return should
+}
+
+// FinishReq implements the Manager interface.
+func (m *managerImpl) FinishReq(g *Guard) {
+	for _, wqg := range g.moveWaitQueueGuards() {
+		m.lt.Dequeue(wqg)
+	}
+	if lg := g.moveLatchGuard(); lg != nil {
+		m.lm.Release(lg)
+	}
+}
+
+// HandleWriterIntentError implements the Manager interface.
+func (m *managerImpl) HandleWriterIntentError(
+	ctx context.Context, g *Guard, t *roachpb.WriteIntentError,
+) (*Guard, *Error) {
+	// Enter or create a txnWaitQueue entry per intent.
+	orig := len(g.wqgs)
+	for _, intent := range t.Intents {
+		g.wqgs = append(g.wqgs, m.lt.AddDiscoveredLock(g.req, intent))
+	}
+	m.lm.Release(g.moveLatchGuard())
+
+	// Wait on each new txnWaitQueue entry.
+	// TODO: should we do this here?
+	for _, wqg := range g.wqgs[orig:] {
+		if err := m.lwq.WaitOn(ctx, g.req, wqg); err != nil {
```

Comment:
This is the case in which I find out about a lock only from the replicated key space, right? And there's the other case in which we discover locks directly from the lock table. I would've expected that we'd wait in `SequenceReq` in all cases, i.e. `HandleWriterIntentError` would be called and after the caller would pass the resulting guard to `SequenceReq`. (This would also explain why my assumption in that method about only having to wait on new wgqs is wrong).

Reply:
Yes, that's exactly how things are supposed to work. I think I just forgot to address this TODO and delete the code.